src/persist-client/src/internal/machine.rs

```diff
@@ -817,22 +849,38 @@ where
 
     async fn apply_unbatched_idempotent_cmd<
         R,
-        WorkFn: FnMut(SeqNo, &PersistConfig, &mut StateCollections<T>) -> ControlFlow<Infallible, R>,
+        WorkFn: FnMut(
+            SeqNo,
+            &PersistConfig,
+            &mut StateCollections<T>,
+        ) -> ControlFlow<NoOpStateTransition<R>, R>,
     >(
         &mut self,
         cmd: &CmdMetrics,
         mut work_fn: WorkFn,
-    ) -> (SeqNo, R, RoutineMaintenance) {
+    ) -> (SeqNo, R, Option<RoutineMaintenance>) {
         let mut retry = self
             .metrics
             .retries
             .idempotent_cmd
             .stream(Retry::persist_defaults(SystemTime::now()).into_retry_stream());
         loop {
             match self.apply_unbatched_cmd(cmd, &mut work_fn).await {
-                Ok((seqno, x, maintenance)) => match x {
-                    Ok(x) => return (seqno, x, maintenance),
-                    Err(infallible) => match infallible {},
+                Ok((seqno, x, mut maintenance)) => match x {
+                    Ok(x) => {
+                        if let Some(tombstone_maintenance) = self.maybe_become_tombstone().await {
+                            maintenance.merge(tombstone_maintenance);
+                        }
+                        return (seqno, x, Some(maintenance));
+                    }
+                    Err(NoOpStateTransition(x)) => {
+                        // apply_unbatched_cmd could also return optional
+                        // maintenance, but various bits of logic are cleaner
+                        // this way. It's always the default (no maintenance)
+                        // for no-ops, so validate that before throwing it away.
```

Comment:
The former! (I originally understood this comment as justifying why we needed to make the distinction here, but on reread I see it's more about explaining why we don't push the distinction any deeper.)

> the former is because that's semantically more accurate. if we NoOp'd out of the cmd application, then it seems to be that we shouldn't be assigned maintenance and this signature communicates that to the caller.

Understood!

I could make the case that `RoutineMaintenance` can already express the "nothing to do" case perfectly well... and it seems like the correct thing to do for a caller in either case is to do any maintenance they are assigned. And I will if that's useful! But mostly I'm just happy to understand the justification.

Reply:
ooh very interesting... I think I like the idea of all calls, even no-ops, receiving maintenance work. it simplifies the signature / expectations a bit, and whether the caller hit a no-op or not, it still tried to participate in changing machine state, so it seems fine to toss some work their way if needed